aws/rust-runtime/aws-sigv4/src/http_request/canonical_request.rs

```diff
@@ -373,11 +409,16 @@ fn trim_spaces_from_byte_string(bytes: &[u8]) -> &[u8] {
     &bytes[starting_index..ending_index]
 }
 
-/// Works just like [trim_all] but acts on HeaderValues instead of bytes
-fn normalize_header_value(header_value: &HeaderValue) -> HeaderValue {
+/// Works just like [trim_all] but acts on HeaderValues instead of bytes.
+/// Will ensure that the underlying bytes are valid UTF-8.
+fn normalize_header_value(header_value: &HeaderValue) -> Result<HeaderValue, Error> {
     let trimmed_value = trim_all(header_value.as_bytes());
-    // This can't fail because we started with a valid HeaderValue and then only trimmed spaces
-    HeaderValue::from_bytes(&trimmed_value).unwrap()
+    let trimmed_value_as_utf8_str =
+        std::str::from_utf8(&trimmed_value).map_err(|e| InvalidHeaderError {
+            source: Error::from(e),
+            kind: InvalidHeaderErrorKind::InvalidUtf8,
+        })?;
+    HeaderValue::from_str(trimmed_value_as_utf8_str).map_err(Error::from)
```

Comment:
Since `Error` is just `Box<dyn Error + Send + Sync + 'static`, I think you can just convert `from_utf8`'s error into it directly to avoid creating `InvalidHeaderError`.

Having the top-level error type be `Box<dyn Error>` is definitely not ideal though. I think that will get addressed through the error refactor though. There's a [RFC for errors](https://github.com/awslabs/smithy-rs/blob/main/design/src/rfcs/rfc0022_error_context_and_compatibility.md) if you want more context on that.

Reply:
Thanks for taking a look!

> Since `Error` is just `Box<dyn Error + Send + Sync + 'static`, I think you can just convert `from_utf8`'s error into it directly to avoid creating `InvalidHeaderError`.

You are right. I try to recall why I did this way and remembered that it was an attempt to incorporate a suggestion by @rcoh
> I think the current error probably won't indicate anything about the problem being a header

You can see that a commit prior to the Russel's comment 1bf2bac did convert `from_uti8`'s error into `Error` directly.

I'm completely fine with removing my half-baked attempt `InvalidHeaderError` and converting `from_uti8`'s error into `Error` directly. Now that I'm going through the mentioned RFC, I think what's described there is a much better way to incorporate the Russell's suggestion moving forward.

What do you think?

Reply:
Agreed. I think we should make a smaller change in this PR, and then fix all the errors in the error refactor.